analysis.ml

```diff
@@ -58,10 +46,111 @@ let forward_analysis (init_state : ('a * pc))
         | Some merged ->
             cell := Some merged;
             let updated = update pc merged in
-            let succs = successors program pc in
+            let succs = successors pc in
             let new_work = List.map (fun pc -> (updated, pc)) succs in
             work (new_work @ rest)
         end
   in
-  work [init_state];
+  work init_state;
   Array.map (!) program_state
+
+let exits program =
+  let rec exits pc : Pc.t list =
+    if Array.length program = pc then []
+    else
+      ((if List.length (successors program pc) = 0 then [pc] else [])
```

Comment:
`successors program pc = []` would be more idiomatic (and also happens to be O(1) even if the list is long.)